### PR TITLE
Add Activate field for the edly-sub-org model

### PR DESCRIPTION
**Pull request by madnihamza1841**
**Description:**
Client Onboarding has been made easy by adding a field in the EdlySubOrganization Model. Middleware has been updated such that, if the field is_active is set to False then the user will be directed to Marketing Page. 

**JIRA:**
https://edlyio.atlassian.net/browse/EDLY-4177

### CODE REVIEW

File: openedx/features/edly/middleware.py [hunk 1]
```diff
@@ -10,7 +10,11 @@
 from django.utils.deprecation import MiddlewareMixin
 
 from openedx.core.djangoapps.site_configuration.models import SiteConfiguration
-from openedx.features.edly.utils import user_has_edly_organization_access
+from openedx.features.edly.utils import (
+    user_has_edly_organization_access,
+    is_edly_sub_org_active,
+    get_marketing_url_from_current_site_configurations,
```
Comment (taimoor-ahmed-1) [inline openedx/features/edly/middleware.py:16]: kindly sort imports alphabetically.

File: openedx/features/edly/models.py [hunk 2]
```diff
@@ -42,6 +42,7 @@ class EdlySubOrganization(TimeStampedModel):
         related_name='edly_sub_org_for_preview_site',
         null=True, on_delete=models.CASCADE
     )
+    is_active = models.BooleanField(default=True, verbose_name=_('is active'))
```
Comment (taimoor-ahmed-1) [inline openedx/features/edly/models.py:45]: it will be good to have some help text for is active field

File: openedx/features/edly/utils.py [hunk 3]
```diff
@@ -26,6 +26,30 @@
...
+    current_site = request.site
+
+    try:
+        edly_sub_org = EdlySubOrganization.objects.get(
+            Q(lms_site=current_site) |
+            Q(studio_site=current_site) |
+            Q(preview_site=current_site)
+        )
+    except EdlySubOrganization.DoesNotExist:
+        return True
```
Comment (taimoor-ahmed-1) [inline openedx/features/edly/utils.py:48]: we should return False here.

File: openedx/features/edly/utils.py [hunk 4]
```diff
@@ -374,6 +398,20 @@ def get_current_site_invalid_certificate_context(default_html_certificate_config
...
+
+    Returns:
+        dict: Context data.
+    """
+    context = dict()
+    current_site_configuration = get_current_site_configuration()
+    if current_site_configuration:
+        context['marketing_url'] = current_site_configuration.get_value('MARKETING_SITE_ROOT')
+
+    return context
```
Comment (taimoor-ahmed-1) [inline openedx/features/edly/utils.py:413]: we can directly return marketing url here as name suggests without adding it in context dict.

File: openedx/features/edly/utils.py [hunk 5]
```diff
@@ -26,6 +26,30 @@
...
+    """
+    Checks if the request EdlySubOrganization is enabled or disabled.
+
+    Arguments:
+        request: HTTP request object
+
+    Returns:
+        bool: Returns True if site is enabled and False if the site is disabled or DoesNotExist
+    """
+    current_site = request.site
```
Comment (taimoor-ahmed-1) [inline openedx/features/edly/utils.py:39]: we can simply pass site here as we don't need complete request object.

**Review by taimoor-ahmed-1 — CHANGES_REQUESTED** [review]

File: openedx/features/edly/utils.py [hunk 6]
```diff
@@ -26,6 +26,20 @@
...
+def is_edly_sub_org_active(edly_sub_org):
+    """
+    Checks if the request EdlySubOrganization is enabled or disabled.
+
+    Arguments:
+        request: EdlySubOrganization model class
+
+    Returns:
+        bool: Returns True if site is enabled and False if the site is disabled or DoesNotExist
+    """
```
Comment (taimoor-ahmed-1) [inline openedx/features/edly/utils.py:38]: if it `DoesNotExist`, this method will break.

File: openedx/features/edly/middleware.py [hunk 7]
```diff
@@ -27,6 +33,26 @@ def process_request(self, request):
         if request.user.is_superuser or request.user.is_staff:
             return
 
+        try:
+            edly_sub_org = EdlySubOrganization.objects.get(
+                Q(lms_site=request.site) |
+                Q(studio_site=request.site) |
+                Q(preview_site=request.site)
+            )
```
Comment (taimoor-ahmed-1) [inline openedx/features/edly/middleware.py:41]: we can avoid querying the db on each request by writing is this way:

```edly_sub_org = None
edly_sub_org = getattr(self.request.site, 'edly_sub_org_for_studio', edly_sub_org)
edly_sub_org = getattr(self.request.site, 'edly_sub_org_for_lms', edly_sub_org)
edly_sub_org = getattr(self.request.site, 'edly_sub_org_for_preview', edly_sub_org)
```

File: openedx/features/edly/utils.py (hunk 6 above)
Comment (madnihamza1841) [reply openedx/features/edly/utils.py]: The function is only called if the EdlySubOrganization exists.

File: openedx/features/edly/utils.py [hunk 8]
```diff
@@ -374,6 +388,19 @@ def get_current_site_invalid_certificate_context(default_html_certificate_config
...
+    Gets the Marketing URL Root Value from Site Configurations
+
+    Returns:
+        dict: Context data.
+    """
+    current_site_configuration = get_current_site_configuration()
+    if current_site_configuration:
+        marketing_url = current_site_configuration.get_value('MARKETING_SITE_ROOT')
+
+    return marketing_url
```
[the remark this answers is not in the document]
Comment (madnihamza1841) [reply openedx/features/edly/utils.py]: Done

File: openedx/features/edly/middleware.py (hunk 7 above)
Comment (madnihamza1841) [reply openedx/features/edly/middleware.py]: Done

**Review by madnihamza1841 — COMMENTED** [review]
All feedback has been addressed.

File: openedx/features/edly/models.py [hunk 9]
```diff
@@ -42,6 +42,11 @@ class EdlySubOrganization(TimeStampedModel):
         related_name='edly_sub_org_for_preview_site',
         null=True, on_delete=models.CASCADE
     )
+    is_active = models.BooleanField(
+        default=True,
+        verbose_name=_('is active'),
```
Comment (burhandodhy) [inline openedx/features/edly/models.py:47]: `'Is Active'`

File: openedx/features/edly/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (taimoor-ahmed-1) [reply openedx/features/edly/utils.py]: if ```if current_site_configuration:``` evaluates to `False`, then `marketing_url` will be undefined and this method will break. Kindly declare `marketing_url` at top of method.

File: openedx/features/edly/middleware.py [hunk 10]
```diff
@@ -27,6 +31,28 @@ def process_request(self, request):
...
+        edly_sub_org = getattr(request.site, 'edly_sub_org_for_lms', edly_sub_org)
+        edly_sub_org = getattr(request.site, 'edly_sub_org_for_preview_site', edly_sub_org)
+
+        if edly_sub_org:
+            if not is_edly_sub_org_active(edly_sub_org):
+                logger.exception('EdlySubOrganization for site %s is disabled. ', request.site)
+                marketing_url = get_marketing_url_from_current_site_configurations()
+
+                if marketing_url:
+                    marketing_disabled_url = marketing_url + "disabled"
```
Comment (taimoor-ahmed-1) [inline openedx/features/edly/middleware.py:44]: kindly use single quotes here.

File: openedx/features/edly/middleware.py [hunk 11]
```diff
@@ -27,6 +31,28 @@ def process_request(self, request):
         if request.user.is_superuser or request.user.is_staff:
             return
 
+        edly_sub_org = getattr(request.site, 'edly_sub_org_for_studio', None)
+        edly_sub_org = getattr(request.site, 'edly_sub_org_for_lms', edly_sub_org)
```
Comment (burhandodhy) [inline openedx/features/edly/middleware.py:35]: We only need the `edly_sub_org_for_lms`. Reason: When the user accesses the studio, Its redirected to LMS for the login same for the preview. In simple if we simple add the check for the LMS, users will not able to loggedIn and eventually it will not able to access our platform.  

cc: @taimoor-ahmed-1

File: openedx/features/edly/models.py (hunk 9 above)
Comment (madnihamza1841) [reply openedx/features/edly/models.py:47]: Done

File: openedx/features/edly/middleware.py (hunk 10 above)
Comment (madnihamza1841) [reply openedx/features/edly/middleware.py:44]: Done.

File: openedx/features/edly/middleware.py (hunk 11 above)
Comment (madnihamza1841) [reply openedx/features/edly/middleware.py]: Done.

File: openedx/features/edly/middleware.py [hunk 12]
```diff
@@ -27,6 +31,25 @@ def process_request(self, request):
         if request.user.is_superuser or request.user.is_staff:
             return
 
+        edly_sub_org = getattr(request.site, 'edly_sub_org_for_lms', None)
+        if edly_sub_org:
+            if not is_edly_sub_org_active(edly_sub_org):
+                logger.exception('EdlySubOrganization for site %s is disabled. ', request.site)
+                marketing_url = get_marketing_url_from_current_site_configurations()
```
Comment (burhandodhy) [inline openedx/features/edly/middleware.py:38]: We need to use the already build function https://github.com/edly-io/edx-platform/blob/develop-juniper/common/djangoapps/edxmako/shortcuts.py#L36

File: openedx/features/edly/middleware.py [hunk 13]
```diff
@@ -27,6 +31,25 @@ def process_request(self, request):
...
+
+                if marketing_url:
+                    marketing_disabled_url = marketing_url + 'disabled'
```
Comment (burhandodhy) [inline openedx/features/edly/middleware.py:41]: The disabled page will be the homepage of the site. We don't need to append anything with the root URL

**Review by burhandodhy — CHANGES_REQUESTED** [review]

File: openedx/features/edly/middleware.py (hunk 12 above)
Comment (madnihamza1841) [reply openedx/features/edly/middleware.py:38]: Dpne

File: openedx/features/edly/middleware.py (hunk 13 above)
Comment (madnihamza1841) [reply openedx/features/edly/middleware.py:41]: Done

**Review by taimoor-ahmed-1 — APPROVED** [review]
LGTM 👍

File: openedx/features/edly/middleware.py [hunk 14]
```diff
@@ -27,6 +31,24 @@ def process_request(self, request):
...
 
+        edly_sub_org = getattr(request.site, 'edly_sub_org_for_lms', None)
+        if edly_sub_org:
+            if not is_edly_sub_org_active(edly_sub_org):
+                logger.exception('EdlySubOrganization for site %s is disabled. ', request.site)
+                marketing_url = marketing_link('ROOT')
+
+                if marketing_url:
+                    return HttpResponseRedirect(marketing_url)
+                else:
```
Comment (burhandodhy) [inline openedx/features/edly/middleware.py:42]: Does this condition get true in any case?

File: openedx/features/edly/middleware.py [hunk 15]
```diff
@@ -27,6 +31,24 @@ def process_request(self, request):
         if request.user.is_superuser or request.user.is_staff:
             return
 
+        edly_sub_org = getattr(request.site, 'edly_sub_org_for_lms', None)
+        if edly_sub_org:
+            if not is_edly_sub_org_active(edly_sub_org):
+                logger.exception('EdlySubOrganization for site %s is disabled. ', request.site)
```
Comment (burhandodhy) [inline openedx/features/edly/middleware.py:37]: Remove extra space from the message

File: openedx/features/edly/tests/test_middleware.py [hunk 16]
```diff
@@ -51,14 +51,27 @@ def setUp(self):
         self.client = Client(SERVER_NAME=self.request.site.domain)
         self.client.login(username=self.user.username, password='test')
 
+    def test_disabbled_edly_sub_orgainzation_access(self):
```
Comment (burhandodhy) [inline openedx/features/edly/tests/test_middleware.py:54]: Add test for the negative case

**Review by burhandodhy — CHANGES_REQUESTED** [review]

File: openedx/features/edly/middleware.py (hunk 14 above)
Comment (madnihamza1841) [reply openedx/features/edly/middleware.py:42]: The marketing_url gets the value from the marketing_link('ROOT') function. 
It will either return the marketing URL if found in the site configurations or else return the default value of '#' if the URL is not found in configurations or settings. 

The condition has been updated to :
                                    if marketing_url != '#':

File: openedx/features/edly/middleware.py (hunk 15 above)
Comment (madnihamza1841) [reply openedx/features/edly/middleware.py:37]: Done.

File: openedx/features/edly/tests/test_middleware.py (hunk 16 above)
Comment (madnihamza1841) [reply openedx/features/edly/tests/test_middleware.py:54]: This is the test for negative case

**Review by burhandodhy — APPROVED** [review]
LGTM

There is typo in `test_disabbled_edly_sub_orgainzation_access` function. Kindly fix it
